### PR TITLE
scoring v1 iOS

### PR DESCRIPTION
**Pull request by emilianojuan**
#### Description:

As part of the contingency plan for the next release, scoring will take place in the app. After downloading the proper configuration file, a score function is used to determine if the user should be notified or not
#### Linked issues:

https://docs.google.com/document/d/1QHjo5TKrMgs6MLtLEj7Kewo8PQ24ukIwK4-AO-t23Zg/edit#

### CODE REVIEW

File: ios/BT/ExposureManager.swift [hunk 1]
```diff
@@ -335,25 +335,33 @@ final class ExposureManager: NSObject {
...
                 return
               }
-              let userExplanation = NSLocalizedString(String.newExposureNotificationBody, comment: .default)
-              self.manager.getExposureInfo(summary: summary!,
-                                               userExplanation: userExplanation) { exposures, error in
-                if let error = error {
-                  self.finish(.failure(error),
+              if let summary = summary, ExposureManager.score(summary: summary, with: configuration) {
+                let userExplanation = NSLocalizedString(String.newExposureNotificationBody, comment: .default)
+                self.manager.getExposureInfo(summary: summary,
```
Comment (lmartensson) [inline ios/BT/ExposureManager.swift:340]: See question to Matt in Slack - I don't know whether we should request getExposureInfo here or not.
We won't be using the result I believe, but this does automatically trigger a notification. Maybe this is how we want to create the notification, maybe it isn't. I think this is fine for now though - it's doing the same thing the existing code does - and if it should be different then that can be done in a follow-up change.

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+extension ExposureConfiguration {
+
+  // Each bucket is capped at 30 minutes, this method calculates what the
+  // weighted duration ends up being capped at.
+
+  var maxWeightedDuration: Int {
+    return Int((attenuationBucketWeights[0] + attenuationBucketWeights[1] + attenuationBucketWeights[3]) * 30 * 60)
```
[the remark this answers is not in the document]
Comment (lmartensson) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:53]: Would be good to add a few simple unit tests to this class, that would've caught the [3] here.

[the remark this answers is not in the document]
Comment (emilianojuan) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:53]: great catch, Lena. I was wondering what tests are needed for this scoring algorithm, like, edge cases and stuff, I am not really aware of what it does so any idea its appreciated

[the remark this answers is not in the document]
Comment (lmartensson) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:53]: Assuming a config with a threshold of 15 minutes, bucket weights 1, 0.5, 0, I would do exposures where:
- matched keys:1, attenuation bucket durations: 900, 0, 0: true. (weighted duration = 15 min because of low bucket)
- matched keys:1, attenuation bucket durations: 800, 0, 0: false. (weighted duration < 15 min)
- matched keys:1, attenuation bucket durations: 0, 900, 0: false. (weighted duration = 7.5 min)
- matched keys:1, attenuation bucket durations: 600, 600, 0: true. (weighted duration = 15 min)
- matched keys:1, attenuation bucket durations: 0, 0, 1800: false. (we don't use this bucket)

-matched keys: 2,  attenuation bucket durations: 900, 0, 0: false. (avg weighted duration = 7.5 min)
-matched keys: 2,  attenuation bucket durations: 900, 1800, 0: true. (avg weighted duration = 15min)

-matched keys: 3,  attenuation bucket durations: 900, 1800, 0: false. (avg weighted duration = 10min)
-matched keys: 3,  attenuation bucket durations: 1800, 1800, 0: true. (avg weighted duration = 15min)

-matched keys: 4, attenuation bucket durations: 900, 1800, 0: false. (max weighted duration not reached)
-matched keys: 4, attenuation bucket durations: 1800, 1800, 0: true. (max weighted duration reached)

[the remark this answers is not in the document]
Comment (emilianojuan) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:53]: this is great, thanks!

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  private static let matchedKeyCountMax = 3
+
+  static func score(summary: ENExposureDetectionSummary,
+                    with configuration: ExposureConfiguration) -> Bool {
+    if summary.matchedKeyCount <= matchedKeyCountMax {
+      // If the weighted average duration per matched key exceeds the
+      // threshold, we know that there was at least one day where the exposure
+      // exceeded the threshold.
+      let matchedKeyCount = Int(summary.matchedKeyCount)
```
[the remark this answers is not in the document]
Comment (emilianojuan) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:14]: no, its an UInt64

File: ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift [hunk 4]
```diff
@@ -8,6 +8,24 @@ import XCTest
 
 // MARK: == Mocks ==
 
+class MockENExposureDetectionSummary: ENExposureDetectionSummary {
```
Comment (mattThousand) [inline ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift:11]: 💯 🧪 👏  (tests 🥇 )

File: ios/BT/ExposureManager.swift [hunk 5]
```diff
@@ -346,18 +341,38 @@ final class ExposureManager: NSObject {
                               completionHandler: completionHandler)
                   return
                 }
-                let newExposures = (exposures ?? []).map { exposure in
-                  Exposure(id: UUID().uuidString,
-                           date: exposure.date.posixRepresentation,
-                           duration: exposure.duration,
-                           totalRiskScore: exposure.totalRiskScore,
-                           transmissionRiskLevel: exposure.transmissionRiskLevel)
+                if let summary = summary, ExposureManager.score(summary: summary, with: configuration) {
```
Comment (johnschoeman) [inline ios/BT/ExposureManager.swift:344]: Noticing this method is getting pretty gnarly, idk what is idiomatic swfit, but should be be pulling some of this out into named helpers/functions?

File: ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift [hunk 6]
```diff
@@ -80,7 +98,7 @@ class APIClientMock: APIClient {
   }
 
   func request<T>(_ request: T, requestType: RequestType, completion: @escaping (Result<JSONObject>) -> Void) where T : APIRequest, T.ResponseType == JSONObject {
-
+    print("ACA??")
```
Comment (johnschoeman) [inline ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift:101]: should this print be removed?

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import Foundation
+import ExposureNotification
+
+extension ExposureManager {
+
+  private static let matchedKeyCountMax = 3
+
+  static func score(summary: ENExposureDetectionSummary,
```
Comment (johnschoeman) [inline ios/BT/Extensions/Exposure Notifications/Scoring.swift:8]:
```suggestion
  static func isAboveScoreThreshold(summary: ENExposureDetectionSummary,
```

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    if matchedKeyCount == 0 {
+      return false
+    }
+    if matchedKeyCount <= matchedKeyCountMax {
+      // If the weighted average duration per matched key exceeds the
+      // threshold, we know that there was at least one day where the exposure
+      // exceeded the threshold.
+      let weightedDurationForKeys = weightedDuration(summary: summary,
+                                              config: configuration) / matchedKeyCount
+      if (weightedDurationForKeys >= configuration.triggerThresholdWeightedDuration * 60) {
```
Comment (johnschoeman) [inline ios/BT/Extensions/Exposure Notifications/Scoring.swift:20]:
```suggestion
     let  isAboveAverageDurationThreshold = (weightedDurationForKeys >= configuration.triggerThresholdWeightedDuration * 60)
     return isAboveAverageDurationThreshold
```

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        return true
+      } else {
+        return false
+      }
+    } else {
+      // If the # matched keys is 4 or more, average duration is no longer
+      // useful because we have reached the cap of the attenuation buckets.
+      // (This assumes a trigger threshold of 15 minutes.)
+      if weightedDuration(summary: summary, config: configuration) >= configuration.maxWeightedDuration {
```
Comment (johnschoeman) [inline ios/BT/Extensions/Exposure Notifications/Scoring.swift:29]:
```suggestion
      let isAboveMaxDurationThreshold = (summary: summary, config: configuration) >= configuration.maxWeightedDuration {
      return isAboveMaxDurationThreshold
```

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift (hunk 8 above)
Comment (johnschoeman) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:20]: not sure if this is valid swift, but the
```
if (expression) 
  return true
else
  return false
```

seems odd to my eye.

File: ios/BT/ExposureManager.swift (hunk 5 above)
Comment (emilianojuan) [reply ios/BT/ExposureManager.swift:344]: yes John, im thinking about adding a promise library to cope with this method complexity, but I feel first we have to get it working correctly with a fair amount of Unit Tests, what do you think?

File: ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift (hunk 6 above)
Comment (emilianojuan) [reply ios/COVIDSafePathsTests/ExposureManagerUnitTests.swift:101]: totally

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift (hunk 8 above)
Comment (emilianojuan) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:20]: yes, its better, thanks

File: ios/BT/Extensions/Exposure Notifications/Scoring.swift (hunk 9 above)
Comment (emilianojuan) [reply ios/BT/Extensions/Exposure Notifications/Scoring.swift:29]: again, better, thanks

File: ios/BT/Extensions/Exposure Notifications/ENManager+Extensions.swift [hunk 10]
```diff
@@ -12,8 +12,12 @@ protocol ExposureNotificationManager {
 
   var exposureNotificationEnabled: Bool { get }
   func setExposureNotificationEnabled(_ enabled: Bool, completionHandler: @escaping ENErrorHandler)
-  func detectExposures(configuration: ENExposureConfiguration, diagnosisKeyURLs: [URL], completionHandler: @escaping ENDetectExposuresHandler) -> Progress
-  @discardableResult func getExposureInfo(summary: ENExposureDetectionSummary, userExplanation: String, completionHandler: @escaping ENGetExposureInfoHandler) -> Progress
+  @discardableResult func detectExposures(configuration: ENExposureConfiguration,
```
Comment (mattThousand) [inline ios/BT/Extensions/Exposure Notifications/ENManager+Extensions.swift:15]: 👍

File: ios/BT/API/Model/DownloadedPackage.swift [hunk 11]
```diff
@@ -2,35 +2,44 @@ import Foundation
...
-  init(keysBin: Data, signature: Data) {
-    bin = keysBin
-    self.signature = signature
-  }
+  static func create(from data: Data) -> Self?
+}
 
-  init?(compressedData: Data) {
-    guard let archive = Archive(data: compressedData, accessMode: .read) else {
+class DownloadedPackage: DownloadableFile {
```
Comment (mattThousand) [inline ios/BT/API/Model/DownloadedPackage.swift:10]: `final` class 🤓

File: ios/BT/API/Model/DownloadedPackage.swift [hunk 12]
```diff
@@ -2,35 +2,44 @@ import Foundation
...
+  init(keysBin: Data, signature: Data) {
+    bin = keysBin
+    self.signature = signature
+  }
+
   let bin: Data
   let signature: Data
 
+  func writeSignatureEntry(toDirectory directory: URL, filename: String) throws -> URL {
+    let url = directory.appendingPathComponent(filename).appendingPathExtension("sig")
```
Comment (mattThousand) [inline ios/BT/API/Model/DownloadedPackage.swift:33]: Not at all necessary, but could use a private extension on `String` for this: 

```
private extension String {
  static let binExtension = "bin"
  static let sigExtension = "sig"
  static let exportFilename = "export"
  static var binEntry: String {
    "\(exportFilename).\(binExtension)"
  }
  static var sigEntry: String {
    "\(exportFilename).\(sigExtension)"
  }
}
```

File: ios/BT/API/Model/DownloadedPackage.swift (hunk 11 above)
Comment (emilianojuan) [reply ios/BT/API/Model/DownloadedPackage.swift:10]: I had to make it a class to be able to inherit a mock class from it, thas why is not final

File: ios/BT/API/Model/DownloadedPackage.swift (hunk 12 above)
Comment (emilianojuan) [reply ios/BT/API/Model/DownloadedPackage.swift:33]: will do
